pkg/capabilities/triggers/mercury_trigger.go

```diff
@@ -225,19 +171,76 @@ func (o *MercuryTriggerService) GetTriggerID(req capabilities.CapabilityRequest,
 	return "", fmt.Errorf("triggerId not found in inputs")
 }
 
-func GenerateTriggerEventID(reports []mercury.FeedReport) string {
-	// Let's hash all the feedIds and timestamps together
-	sort.Slice(reports, func(i, j int) bool {
-		if reports[i].FeedID == reports[j].FeedID {
-			return reports[i].ObservationTimestamp < reports[j].ObservationTimestamp
+func (o *MercuryTriggerService) loop() {
+	defer o.wg.Done()
+	now := time.Now().UnixMilli()
+	nextWait := o.tickerResolutionMs - now%o.tickerResolutionMs
+
+	for {
+		select {
+		case <-o.stopCh:
+			return
+		case <-time.After(time.Duration(nextWait) * time.Millisecond):
+			startTs := time.Now().UnixMilli()
+			// find closest timestamp that is a multiple of o.tickerResolutionMs
+			aligned := (startTs + o.tickerResolutionMs/2) / o.tickerResolutionMs * o.tickerResolutionMs
+			o.process(aligned)
+			currTs := time.Now().UnixMilli()
+			desiredNext := aligned + o.tickerResolutionMs
+			nextWait = desiredNext - currTs
```

Comment:
I think having this as a private function and explicitly table tested would be good. It's hard as a reviewer to see exactly what's happening here. Its a good candidate for unit testing